src/history/hs.c

```diff
@@ -691,67 +711,49 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
          * tombstone.
          * 4) We have a single tombstone on the chain, it is simply ignored.
          */
-        for (; upd != NULL; upd = upd->next) {
+        for (prev_upd = NULL; upd != NULL; prev_upd = upd, upd = upd->next) {
             if (upd->txnid == WT_TXN_ABORTED)
                 continue;
+
             WT_ERR(__wt_modify_vector_push(&modifies, upd));
 
             /*
-             * If the update is the second update older than the prepared update and we haven't seen
-             * a tombstone. Mark the update.
+             * Only insert full update to the history store if we write a prepared update to the
+             * data store.
              */
-            if (upd->prepare_state == WT_PREPARE_INPROGRESS) {
-                /*
-                 * No normal update between prepared updates and the first prepared update cannot be
-                 * a tombstone.
-                 */
-                WT_ASSERT(session, (track_prepare && upd_count == 0) ||
-                    (!track_prepare && upd->type != WT_UPDATE_TOMBSTONE));
-                track_prepare = true;
-            } else if (track_prepare) {
-                if (upd->type == WT_UPDATE_TOMBSTONE) {
-                    upd_count = 0;
-                    track_prepare = false;
-                } else if (upd_count == 0)
-                    ++upd_count;
-                else {
-                    second_older_than_prepare = upd;
-                    upd_count = 0;
-                    track_prepare = false;
-                }
-            }
+            if (upd->prepare_state == WT_PREPARE_INPROGRESS)
+                enable_reverse_modify = false;
+
+            /* Only insert full update to the history store if we need to squash the updates. */
+            if (prev_upd != NULL && prev_upd->txnid == upd->txnid &&
+              prev_upd->start_ts == upd->start_ts)
```

Comment:
There will be cases where a single transaction does multiple updates with different timestamps. That happens on replica set secondaries. Is that OK? i.e: is it safe to save a modify whenever the timestamp differs, even if the transaction ID is the same?

Reply:
@agorrod I believe that is OK as long as the timestamps are in order.

Reply:
Sounds good.